static/graph-layout-core.ts

```diff
@@ -0,0 +1,849 @@
+        // This doesn't matter too much as far as readability goes
+        //
+        //      A            A
+        //     / \          / \
+        //    B   C   or   B   C
+        //    |                |
+        //    D                D
+        for (const i of topologicalOrder) {
+            // Only dag edges are considered
+            // Edges - dag edges = the set of back edges
+            const block = this.blocks[i];
+            for (const j of block.dagEdges) {
+                const target = this.blocks[j];
+                if (target.treeParent === null && target.row === block.row + 1) {
+                    block.treeEdges.push(j);
+                    target.treeParent = i;
+                }
+            }
+        }
+    }
+
+    adjustSubtree(root: number, rowShift: number, columnShift: number) {
+        const block = this.blocks[root];
+        block.row += rowShift;
+        block.col += columnShift;
+        for (const j of block.treeEdges) {
+            this.adjustSubtree(j, rowShift, columnShift);
+        }
+    }
+
+    // Note: Currently O(n^2)
+    computeTreeColumnPositions(node: number) {
+        const block = this.blocks[node];
+        if (block.treeEdges.length === 0) {
+            block.row = 0;
+            block.col = 0;
+            block.boundingBox = {
+                rows: 1,
+                cols: 2,
+            };
+        } else if (block.treeEdges.length === 1) {
+            const childIndex = block.treeEdges[0];
+            const child = this.blocks[childIndex];
+            block.row = 0;
+            block.col = child.col;
+            block.boundingBox = {
+                rows: 1 + child.boundingBox.rows,
+                cols: child.boundingBox.cols,
+            };
+            this.adjustSubtree(childIndex, 1, 0);
+        } else {
+            // If the node has more than two children we'll just center between the
+            //let selectedTreeEdges = block.treeEdges.slice(0, 2);
+            const boundingBox = {
+                rows: 0,
+                cols: 0,
+            };
+            // Compute bounding box of all the subtrees and adjust
+            for (const i of block.treeEdges) {
+                const child = this.blocks[i];
+                this.adjustSubtree(i, 1, boundingBox.cols);
+                boundingBox.rows += child.boundingBox.rows;
+                boundingBox.cols += child.boundingBox.cols;
+            }
+            // Position parent
+            boundingBox.rows++;
+            block.boundingBox = boundingBox;
+            block.row = 0;
+            // between immediate children
+            const [left, right] = [this.blocks[block.treeEdges[0]], this.blocks[block.treeEdges[1]]];
+            block.col = Math.floor((left.col + right.col) / 2); // TODO
+        }
+    }
+
+    assignColumns(topologicalOrder) {
+        // Note: Currently not taking shape into account like Cutter does.
```

Comment:
Skipping full shape based side by side placement this might be fine. Typical function size in compiler explorer probably won't be as big as when doing reverse engineering.  Also with the compacting step (which you currently don't have) fullshape  based placement is somewhat redundant.  So it might be better to start with that. Although that's also a big nasty piece of code, maybe less if you used of the shelf LP solver.